app/assets/javascripts/libs/diffable_map.js

```diff
@@ -0,0 +1,294 @@
+// @flow
+// import Utils from "libs/utils";
+
+const defaultItemsPerBatch = 10000;
+let idCounter = 0;
+
+// Some properties:
+// Insertion order is not guaranteed
+
+class DiffableMap<K: number, V> {
+  id: number;
+  chunks: Array<Map<K, V>>;
+  entryCount: number;
+  existsCache: Map<K, boolean>;
+  itemsPerBatch: number;
+
+  constructor(optKeyValueArray?: ?Array<[K, V]>, itemsPerBatch: number = defaultItemsPerBatch) {
+    // Make the id property not enumerable so that it does not interfere with tests
+    Object.defineProperty(this, "id", { value: idCounter++, writable: true });
+    // this.id = idCounter++;
+    this.chunks = [];
+    this.existsCache = new Map();
+    this.entryCount = 0;
+    this.itemsPerBatch = itemsPerBatch;
+
+    if (optKeyValueArray != null) {
+      for (const [key, value] of optKeyValueArray) {
+        this.mutableSet(key, value);
+      }
+    }
+  }
+
+  get(key: K): V {
+    let idx = 0;
+    while (this.chunks[idx] != null) {
+      if (this.chunks[idx].has(key)) {
+        // $FlowFixMe
+        return this.chunks[idx].get(key);
+      }
+      idx++;
+    }
+    throw new Error("Get empty");
+  }
+
+  has(key: K): boolean {
+    try {
+      this.get(key);
+      return true;
+    } catch (exception) {
+      return false;
+    }
+  }
+
+  set(key: K, value: V): DiffableMap<K, V> {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          const newMap = shallowCopy(this);
```

Comment:
Why do we create new map for existing keys? Just for the sake of `immutability`? How does V8 / JS handle immutability and `Map` objects?

Reply:
Yes, changing the map should always be done in a new instance to achieve immutability. V8 was quite fast at copying Map instances, when I tested it.